### PR TITLE
INGEST: Fix ThreadWatchDog Throwing on Shutdown

### LINKED ISSUE #32539: [CI] o.e.i.c.IngestRestartIT.testScriptDisabled fails with EsRejectedExecutionException
`o.e.i.c.IngestRestartIT.testScriptDisabled` failed as follows, apparently due to trying to use a threadpool after it had terminated. Full log at https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/2443/console

```
  1> [2018-08-01T05:50:12,899][INFO ][o.e.i.c.IngestRestartIT  ] [testScriptDisabled] [IngestRestartIT#testScriptDisabled]: cleaned up after test
  1> [2018-08-01T05:50:12,899][INFO ][o.e.i.c.IngestRestartIT  ] [testScriptDisabled] after test
ERROR   4.05s | IngestRestartIT.testScriptDisabled <<< FAILURES!
   > Throwable #1: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=66, name=elasticsearch[node_t0][generic][T#2], state=RUNNABLE, group=TGRP-IngestRestartIT]
   >  at __randomizedtesting.SeedInfo.seed([F226344894683122:B5BAEEDA56780614]:0)
   > Caused by: org.elasticsearch.common.util.concurrent.EsRejectedExecutionException: rejected execution of java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@60e56eae on java.util.concurrent.ScheduledThreadPoolExecutor@2a8207d9[Terminated, pool size = 1, active threads = 0, queued tasks = 0, completed tasks = 2]
   >  at __randomizedtesting.SeedInfo.seed([F226344894683122]:0)
   >  at org.elasticsearch.common.util.concurrent.EsAbortPolicy.rejectedExecution(EsAbortPolicy.java:48)
   >  at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
   >  at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:326)
   >  at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:533)
   >  at org.elasticsearch.threadpool.ThreadPool.schedule(ThreadPool.java:346)
   >  at org.elasticsearch.ingest.IngestService.lambda$new$0(IngestService.java:49)
   >  at org.elasticsearch.grok.ThreadWatchdog$Default.interruptLongRunningExecutions(ThreadWatchdog.java:143)
   >  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:624)
   >  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   >  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   >  at java.lang.Thread.run(Thread.java:748)
```

I see a similar failure on 13 June but nothing else:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+periodic/2179/console

The REPRODUCE WITH line does not reproduce, but is here for posterity:

```./gradlew :modules:ingest-common:integTestRunner -Dtests.seed=F226344894683122 -Dtests.class=org.elasticsearch.ingest.common.IngestRestartIT -Dtests.method="testScriptDisabled" -Dtests.security.manager=true -Dtests.locale=ar-SY -Dtests.timezone=Africa/Banjul```

### PR DESCRIPTION
**Pull request by original-brownbear**
* #32539 is caused by the fact that ThreadWatchDog.Default could throw on shutdown if the ThreadPool is interrupted while `interruptLongRunningExecutions` is in progress. This is a result of the watchdog not having a lifecycle of its own (normally it terminates when the threadpool terminates).
  * We can't easily use `org.elasticsearch.common.util.concurrent.EsRejectedExecutionException#isExecutorShutdown` to catch this state the same way other components do since thatwould require adding the core lib to Grok as a dependency
  * Since we have no knowledge of the lifecycle in this compontent since we're only passed the scheduler `BiFunction` I fixed this by only scheduling the watchdog when there's actually registered threads in it.
    * I think using the patter of locking via two `Atomic*` values should not be much of a performance concern here under load since either the integer will be > 0 in this case (because we have multiple Grok in parallel) or the running state will be true because there likely was at least one thread registered when the watchdog ran and so the enqueing of the watchdog task during `register` will happen very rarely here. The atomic update on the count should not be relevant relative to the cost of adding a new node to the CHM either.
* Fixes #32539
  * Also fixes the watchdog to run if it doens't have to in general.

### CODE REVIEW

File: libs/grok/src/main/java/org/elasticsearch/grok/ThreadWatchdog.java [hunk 1]
```diff
@@ -114,11 +118,14 @@ private Default(long interval,
             this.maxExecutionTime = maxExecutionTime;
             this.relativeTimeSupplier = relativeTimeSupplier;
             this.scheduler = scheduler;
-            scheduler.apply(interval, this::interruptLongRunningExecutions);
         }
         
         public void register() {
+            registered.getAndIncrement();
             Long previousValue = registry.put(Thread.currentThread(), relativeTimeSupplier.getAsLong());
+            if (running.compareAndSet(false, true) == true) {
```
[the remark this answers is not in the document]
Comment (talevy) [reply libs/grok/src/main/java/org/elasticsearch/grok/ThreadWatchdog.java]: also, can't `running` be set to false from one grok-processor execution in one thread at any point after this compareAndSet is run? So this wouldn't technically be a correct boolean value at all times.

[the remark this answers is not in the document]
Comment (original-brownbear) [reply libs/grok/src/main/java/org/elasticsearch/grok/ThreadWatchdog.java]: @talevy the only way `false` is set, is when there's no single Grok action running while `interruptLongRunningExecutions` is executed, there is no other code that sets `false` for this variable.
Functionally this change shouldn't have any effect on the timeout enforcement timings.

[the remark this answers is not in the document]
Comment (talevy) [reply libs/grok/src/main/java/org/elasticsearch/grok/ThreadWatchdog.java]: OK. Do you see a way to add unit tests for the modified registers and unregisters?

**Comment by original-brownbear** [comment]
@talevy thanks for taking a look! Added a test now :)

**Comment by original-brownbear** [comment]
@talevy ping :)

**Review by talevy — APPROVED** [review]
thanks!

**Comment by original-brownbear** [comment]
@talevy thanks for the review!

**Pull request merged**